nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/util/SCPTransfer.java

```diff
@@ -0,0 +1,682 @@
+            } catch (SftpException e) {
+                if (e.id != ChannelSftp.SSH_FX_FAILURE) {
+                    throw new IOException("Could not blindly create remote directory due to " + e.getMessage(), e);
+                }
+            }
+            return;
+        }
+        // end if disableDirectoryListing
+
+        boolean exists;
+        try {
+            channel.stat(remoteDirectory);
+            exists = true;
+        } catch (final SftpException e) {
+            if (e.id == ChannelSftp.SSH_FX_NO_SUCH_FILE) {
+                // No Such File
+                exists = false;
+            } else {
+                throw new IOException("Failed to determine if remote directory exists at " + remoteDirectory + " due to " + e, e);
+            }
+        }
+
+        if (!exists) {
+            // first ensure parent directories exist before creating this one
+            if (directoryName.getParent() != null && !directoryName.getParentFile().equals(new File(File.separator))) {
+                ensureDirectoryExists(flowFile, directoryName.getParentFile());
+            }
+            logger.debug("Remote Directory {} does not exist; creating it", new Object[] {remoteDirectory});
+            try {
+                channel.mkdir(remoteDirectory);
+                logger.debug("Created {}", new Object[] {remoteDirectory});
+            } catch (final SftpException e) {
+                throw new IOException("Failed to create remote directory " + remoteDirectory + " due to " + e, e);
+            }
+        }
+    }
+
+    private ChannelSftp getChannel(final FlowFile flowFile) throws IOException {
+        if (sftp != null) {
+            String sessionhost = session.getHost();
+            String desthost = ctx.getProperty(HOSTNAME).evaluateAttributeExpressions(flowFile).getValue();
+            if (sessionhost.equals(desthost)) {
+                // destination matches so we can keep our current session
+                return sftp;
+            } else {
+                // this flowFile is going to a different destination, reset session
+                close();
+            }
+        }
+
+        final JSch jsch = new JSch();
+        try {
+            final Session session = jsch.getSession(ctx.getProperty(USERNAME).evaluateAttributeExpressions(flowFile).getValue(),
+                ctx.getProperty(HOSTNAME).evaluateAttributeExpressions(flowFile).getValue(),
+                ctx.getProperty(PORT).evaluateAttributeExpressions(flowFile).asInteger().intValue());
+
+            final String hostKeyVal = ctx.getProperty(HOST_KEY_FILE).getValue();
+            if (hostKeyVal != null) {
+                jsch.setKnownHosts(hostKeyVal);
+            }
+
+            final Properties properties = new Properties();
+            properties.setProperty("StrictHostKeyChecking", ctx.getProperty(STRICT_HOST_KEY_CHECKING).asBoolean() ? "yes" : "no");
+            properties.setProperty("PreferredAuthentications", "publickey,password");
```

Comment:
Agreed. But i think we should raise  separate JIRA for SFTPTransfer fix and don't munge the two together. That said I am also seeing an opportunity for some refactoring as the code you are pointing to almost identical (if not identical) to the proposed SCP code and could easily be encapsulated into some abstract/utility class used by both to avoid code duplication. Not sure how to proceed with it properly. . .